src/libs/sfdk/dockervirtualmachine.cpp

```diff
@@ -0,0 +1,529 @@
+{
+    Q_UNUSED(ruleName)
+    Q_UNUSED(context)
+    Q_UNUSED(functor)
+    QTC_CHECK(false);
+}
+
+void DockerVirtualMachinePrivate::doSetReservedPortForwarding(
+        VirtualMachinePrivate::ReservedPort which, quint16 port, const QObject *context,
+        const Functor<bool> &functor)
+{
+    buildWithLabel(labelFor(which), QString::number(port), context, functor);
+}
+
+void DockerVirtualMachinePrivate::doSetReservedPortListForwarding(
+        VirtualMachinePrivate::ReservedPortList which, const QList<Utils::Port> &ports,
+        const QObject *context, const Functor<const QMap<QString, quint16> &, bool> &functor)
+{
+    Q_UNUSED(which)
+    Q_UNUSED(ports)
+    Q_UNUSED(context)
+    Q_UNUSED(functor)
+    QTC_CHECK(false);
+}
+
+void DockerVirtualMachinePrivate::doRestoreSnapshot(const QString &snapshotName,
+        const QObject *context, const Functor<bool> &functor)
+{
+    Q_UNUSED(snapshotName)
+    Q_UNUSED(context)
+    Q_UNUSED(functor)
+    QTC_CHECK(false);
+}
+
+VirtualMachineInfo DockerVirtualMachinePrivate::virtualMachineInfoFromOutput(
+        const QByteArray &output) const
+{
+    QJsonDocument document = QJsonDocument::fromJson(output);
+    VirtualMachineInfo info = {};
+
+    QTC_ASSERT(!document.isNull(), return info);
+
+    QJsonObject labels = document.object();
+
+    auto parsePort = [=](VirtualMachinePrivate::ReservedPort which, quint16 *port) {
+        QString key = labelFor(which);
+        if (labels.contains(key))
+            *port = labels.value(key).toString().toUShort();
+    };
+
+    parsePort(VirtualMachinePrivate::SshPort, &info.sshPort);
+    parsePort(VirtualMachinePrivate::WwwPort, &info.wwwPort);
+
+    auto parsePath = [=](VirtualMachinePrivate::SharedPath which, QString *path) {
+        QString key = labelFor(which);
+        if (labels.contains(key))
+            *path = labels.value(key).toString();
+    };
+
+    parsePath(VirtualMachinePrivate::SharedSrc, &info.sharedSrc);
+    parsePath(VirtualMachinePrivate::SharedSsh, &info.sharedSsh);
+    parsePath(VirtualMachinePrivate::SharedHome, &info.sharedHome);
+    parsePath(VirtualMachinePrivate::SharedConfig, &info.sharedConfig);
+    parsePath(VirtualMachinePrivate::SharedTargets, &info.sharedTargets);
+
+    info.memorySizeMb = VirtualMachine::availableMemorySizeMb();
+    info.cpuCount = VirtualMachine::availableCpuCount();
+
+    return info;
+}
+
+QString DockerVirtualMachinePrivate::labelFor(VirtualMachinePrivate::SharedPath which)
+{
+    QString label;
+    switch (which) {
+    case SharedSrc:
+        label = Constants::BUILD_ENGINE_SHARED_SRC;
+        break;
+    case SharedHome:
+        label = Constants::BUILD_ENGINE_SHARED_HOME;
+        break;
+    case SharedConfig:
+        label = Constants::BUILD_ENGINE_SHARED_CONFIG;
+        break;
+    case SharedSsh:
+        label = Constants::BUILD_ENGINE_SHARED_SSH;
+        break;
+    case SharedTargets:
+        label = Constants::BUILD_ENGINE_SHARED_TARGET;
+    break;
```

Comment:
bad indent.

Better return immediately from the cases?

Reply:
fixed